### ISSUE TITLE
Minimal example of using your package in conjunction with a foreach loop and the doparallel backend

### ISSUE THREAD

**Issue by Hallo951** [mihaiconstantin/parabar#28]
Hello,

I came across your package while searching for a progressbar for the foreach loop with the doparallel backend. 

So far I have always used the dosnow parallel backend with the dosnow progressbar. Since I've been having problems with the dosnow backend in parallel executions lately (unserial error -> worker death), I switched to the doparallel backend, with which these problems no longer occurred.  However, I now found that with the doparallel backend, the dosnow progressbar no longer works (no progress is displayed). 

Can you post a minimal example of using your package in conjunction with a foreach loop and the doparallel backend here? That would be very helpful.

Thank you very much.

best Frank

**Comment by Hallo951** [comment]
Or did I understand correctly that your package is based on the doparall backend and if I want to have a working progressbar with foreach loops, I have to replace the doparall backend with your backend? In this case, do I still need the dosnow option in the foreach loop? A minimal example would be really good.

**Comment by mihaiconstantin** [comment]
Hi Frank,

I am not familiar with the `foreach` and `doParallel` packages. Knowing the design of `parabar`, they probably can be incorporated. However, with `parabar`, the goal was two-fold:

1. Stick to the built-in `parallel` package in `R` for creating clusters and managing the execution of parallel tasks. 
    - I wanted to stay as close as possible to the built-in capabilities and reduce external dependencies (e.g., only using `R6`, `progress`, `callr`, and `filelock`).

2. Provide a standard interface to work with `parallel` clusters.
    - Essentially, on `parallel` clusters you can perform operations (e.g., exporting variables, evaluating expressions, etc.), and these are defined by some sort of an interface as follows:
        - `start`: Start the backend.
        - `stop`: Stop the backend.
        - `clear`: Remove all objects from the backend.
        - `peek`: Show the variables names available on the backend.
        - `export`: Export variables from a given environment to the backend.
        - `evaluate`: Evaluate an arbitrary expression on the backend.
        - `sapply`: Run a task on the backend.
        - `get_output`: Get the output of the task execution.

Then, `parabar` allows you to *decorate* any of these operations to change their behavior. For example, the `sapply` operation can be decorated to track the progress of tasks and display a progress bar. Here is a [more detailed discussion](https://parabar.mihaiconstantin.com/articles/comparison.html) on the specific progress bar functionality.

While examples are also discussed on the website (i.e., [parabar.mihaiconstantin.com](https://parabar.mihaiconstantin.com)), below you can find a condensed version.

### User API
This API is meant for use in interactive scripts.

```r
# Define your task function.
task <- function(x) {
    # Sleep a bit.
    Sys.sleep(0.01)

    # Compute and return.
    x + 1
}

# Start a synchronous backend.
backend <- start_backend(cores = 4, cluster_type = "psock", backend_type = "async")

# Run the task in parallel.
results <- par_sapply(backend, x = 1:1000, fun = task)

# Change the progress bar options.
configure_bar(type = "modern", format = "[:bar] :percent")

# Run the task in parallel.
results <- par_sapply(backend, x = 1:1000, fun = task)

# Stop the backend.
stop_backend(backend)
```

### Developer API
This API is more flexible and meant to be used in other packages.

```r
# Create a specification object.
specification <- Specification$new()
specification$set_cores(4)
specification$set_type("psock")

# Create a backend factory.
backend_factory <- BackendFactory$new()

# Get an asynchronous backend instance.
backend <- backend_factory$get("async")

# Start the backend.
backend$start(specification)

# Create a context factory.
context_factory <- ContextFactory$new()

# Get a progress-tracking context.
context <- context_factory$get("progress")

# Register the backend with the context.
context$set_backend(backend)

# Create a bar factory.
bar_factory <- BarFactory$new()

# Get a `modern` bar (i.e., via `progress::progress_bar`).
bar <- bar_factory$get("modern")

# Register the `bar` with the `context`.
context$set_bar(bar)

# Configure the `bar`.
context$configure_bar(
    show_after = 0,
    format = " > completed :current out of :total tasks [:percent] [:elapsed]"
)

# Run the task in parallel with progress tracking.
context$sapply(x = 1:1000, fun = task)

# Get the results.
results <- context$get_output(wait = TRUE)

# Configure the `bar`.
context$configure_bar(
    show_after = 0,
    format = "[:bar] :percent"
)

# Run the task in parallel with progress tracking.
context$sapply(x = 1:1000, fun = task)

# Get the results.
results <- context$get_output(wait = TRUE)

# Stop the backend.
context$stop()
```    

Both APIs essentially do the same thing. The class-based one just gives you more control on how to create your own backends, contexts, and bars. This might not be the answer you are looking for, but I hope it helps a bit nevertheless.

Also, note that currently only the `sapply` operation is implemented, but implementing others (e.g., `apply`, `lapply`) is trivial and I am open do adding them if there is interest.

**Comment by Hallo951** [comment]
Thanks for your answer. I became aware of the package via a link on [this page](https://stackoverflow.com/questions/66604588/showing-progress-bar-with-doparallel-foreach). I thought that your package therefore also supported foreach loops. Do you see a way to integrate your package into the foreach environment?

**Comment by mihaiconstantin** [comment]
> Thanks for your answer. I became aware of the package via a link on [this page](https://stackoverflow.com/questions/66604588/showing-progress-bar-with-doparallel-foreach). I thought that your package therefore also supported foreach loops.

Ah, I remember that post. I was looking for a way to show progress bars for my parallel tasks and couldn't find anything reliable. I recall that one solution I found was not accurately reflecting the progress of the tasks. That's what lead to the development of `parabar`, but in the process I decided to stick to `R`'s built-in functionality.

> Do you see a way to integrate your package into the foreach environment?

It certainly is possible. When I have some time, I will take a look at the `foreach` operators to see how they work. But just for my own understanding--what is the advantage of using `foreach` syntax over, say, `sapply`?

**Comment by Hallo951** [comment]
That is a good question. Maybe sapply will work too. Here are some examples of how I use foreach. Maybe you can tell me the sapply alternatives:


```
parallel <- T # or F
# switch parallel/ non parallel
 `%op%` <- ifelse(parallel,`%dopar%`,`%do%`)
              
tmp <- 
foreach(a = iter(x, by='column'), .packages = c("nortest","rcompanion"), .combine = cbind) %op% {
       # do something
       # return a data.frame
}
```

```
silent <- T  # or F

  # dosnow progressbar
  pb <- if(!silent){txtProgressBar(max = length(model.formulas), style = 3)}else{NULL}
  
  tmp_all <-
    foreach(j=model.formulas # list of model formulas
            ,.packages = c("spTimer","stats","foreach")
            ,.combine = rbind
            ,.options.snow = if(!silent){opts}else{NULL}) %dopar% {
                          # do something
                          # return a data.frame
            }
```


```
silent <- T  # or F

# initial Progressbar for foreach loop
    pb <- if(!silent){txtProgressBar(max = (nrow(tmp_all) * length(lambda_seq) * length(data.cross)), style = 3)}else{NULL}
    
    # calculate truncated GPP models with different knoth grids and k-fold cross validated
    tmp_cv <-
      foreach(mod = iter(tmp_all, by = "row"),
              .packages = c("spTimer"), .options.snow = if(!silent){opts}else{NULL}, .combine = function(x,...){as.data.frame(rbind(x,...))}) %:%
      foreach(lambda = lambda_seq, 
              .packages = c("spTimer"), .combine = rbind) %:%
      foreach(cv_folder = data.cross,
              .packages = c("spTimer"), .combine = function(x,...){colMeans(rbind(x,...),na.rm = T)}) %dopar% {
# do something
# return a data.frame
}
```


```
silent <- T  # or F

      # initial Progressbar for foreach loop
      pb <- if(!silent){txtProgressBar(max = (length(test_formulas$formula) * length(data.cross)), style = 3)}else{NULL}

      # calculate truncated models
      tmp <- 
        foreach(mod = iter(test_formulas, by = "row"),
                .packages = c("spTimer"), .options.snow = if(!silent){opts}else{NULL}, .combine = function(x,...){as.data.frame(rbind(x,...))}) %:%
          foreach(cv_folder = data.cross,
                  .packages = c("spTimer"), .combine = function(x,...){colMeans(rbind(x,...),na.rm = T)}) %dopar% {
# do something
# return a data.frame
}
```

```
formulas <- unlist(foreach(j=single.formulas, .packages = "rlist") %dopar% {
# do something
}
```

**Comment by mihaiconstantin** [comment]
Closing as duplicate of #29 and #30.
